pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +173,137 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
+		}
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(graceCondition, oldGraceCondition))
+
+		var err error
+		if requeueTime != 0 {
+			degradedCondition.Status = operatorv1.ConditionFalse
+			degradedCondition.Reason = "Waiting"
+			degradedCondition.Message = graceCondition.Message
+			err = retry.New(errors.New("DNS service is waiting: "+graceCondition.Message), requeueTime)
+		} else {
+			// If not requeueing, then mark this as degraded, despite the status of the grace condition
+			degradedCondition.Status = operatorv1.ConditionTrue
+			degradedCondition.Reason = "GracePeriodExpired"
+			degradedCondition.Message = graceCondition.Message
+		}
+		degradedConditions = append(degradedConditions, setDNSLastTransitionTime(degradedCondition, oldDegradedCondition))
+
+		return degradedConditions, err
+	}
+
+	// nothing is degraded, with or without a grace condition
+	degradedCondition.Status = operatorv1.ConditionFalse
+	degradedCondition.Reason = "AsExpected"
+	degradedCondition.Message = "Enough DNS pods are available, and the DNS service has a cluster IP address."
+
+	return append(degradedConditions, setDNSLastTransitionTime(degradedCondition, oldDegradedCondition)), nil
+}
+
+// requeueAfter compares expected operator conditions to the grace
+// condition and returns a requeueing wait time.
+func requeueAfter(expectedConds []expectedCondition, graceCondition, oldGraceCondition *operatorv1.OperatorCondition) time.Duration {
+	var requeueAfter time.Duration
+	now := clock.Now()
+	init := true
+	var t2 time.Time
+	if oldGraceCondition != nil {
+		t2 = oldGraceCondition.LastTransitionTime.Time
 	} else {
-		degradedCondition.Status = operatorv1.ConditionFalse
-		degradedCondition.Reason = "AsExpected"
-		degradedCondition.Message = "Enough DNS pods are available, and the DNS service has a cluster IP address."
+		t2 = now
 	}
+	conditionsMap := make(map[string]*operatorv1.OperatorCondition)
+	conditionsMap[graceCondition.Type] = graceCondition
+
+	for _, expected := range expectedConds {
+		condition, haveCondition := conditionsMap[expected.condition]
+		if !haveCondition {
+			continue
+		}
+		if condition.Status == expected.status {
+			continue
+		}
+		if expected.gracePeriod != 0 {
+			t1 := now.Add(-expected.gracePeriod)
+			if t2.After(t1) {
+				d := t2.Sub(t1)
+				if init || d < requeueAfter {
+					// Set the requeueAfter value
+					// if the grace period hasn't elapsed
+					requeueAfter = d
+					init = false
+				}
+				continue
+			}
+		}
+	}
+	return requeueAfter
+}
+
+func formatConditions(conditions []operatorv1.OperatorCondition) (string, string) {
+	var reasons, messages string
+	if len(conditions) == 0 {
+		return "", ""
+	}
+	for _, cond := range conditions {
+		reasons = reasons + fmt.Sprintf(", %s=%s (%s)", cond.Type, cond.Status, cond.Reason)
+		messages = messages + fmt.Sprintf(", %s=%s (%s)", cond.Type, cond.Status, cond.Message)
+	}
+	return reasons[2:], messages[2:]
```

Comment:
The status condition's reason should be a CamelCase string.  You could do something like this:
```suggestion
		reasons = reasons + "And" + cond.Reason
		messages = messages + fmt.Sprintf(", %s=%s (%s)", cond.Type, cond.Status, cond.Message)
	}
	return reasons[3:], messages[2:]
```
Or just use something like "DegradedConditions" as the reason.